Clarify limits of submit --bundle --parallel (#157)
glotzerlab/signac-docs: docs/source/cluster_submission.rst

Bundling
========

By default, all eligible job-operations will be submitted as separate cluster jobs.
This is usually the best model for clusters that provide shared compute partitions because it allows the cluster scheduler to optimize the scheduling of your job.
However, sometimes it is beneficial to execute multiple operations within one cluster job, like if the compute cluster can only make reservation for full nodes or if there is a limit to the number of cluster jobs you can submit to the cluster scheduler's queue.

You can execute multiple job-operations in serial per cluster job submission with the ``--bundle`` option.
For example, the following command will bundle up to five job-operations to be executed in each cluster submission:

.. code-block:: bash

    ~/my_project $ python project.py submit --bundle=5

Without any argument, the ``--bundle`` option will bundle **all** eligible job-operations into a single cluster job.

.. tip::

    Recognizing that ``--bundle=1`` is the default option might help you to better understand the bundling concept.

By default, the submit command will run bundled job-operations in serial.

> It is possible to run multiple CPU processes in parallel, as long as the operating system assigns running threads to any available CPU cores, which is not the default behavior on all compute clusters.

Reviewer: ```suggestion
It is possible to request bundled CPU jobs run in parallel with the ``--parallel`` flag, as long as the operating system assigns running threads to any available CPU cores, which is NOT the default behavior on all compute clusters.
```

Reviewer 2: ```suggestion
It is possible to run bundled jobs in parallel (when executing on the CPU), as long as the operating system assigns running threads to any available CPU cores, which is **not** the default behavior on all compute clusters.
```

The ``--parallel`` flag will start the bundled job-operations and run them as background processes. Ensure that the processes are correctly assigned to the requested resources before using this option.

.. warning::

>     The ``--parallel`` option will not distribute operations among multiple GPUs. Use aggregation instead. (TODO: link to tutorial)

Reviewer 3: This TODO must be resolved before merging.

    Use the ``--pretend`` option to preview the generated submission scripts on screen instead of submitting them.

Bundling
========

By default, all eligible job-operations will be submitted as separate cluster jobs.

> This is usually the best model for clusters that provide shared compute partitions
> because it allows the cluster scheduler to optimize the scheduling of your job.

Reviewer 3: One sentence per line.
```suggestion
This is usually the best model for clusters that provide shared compute partitions because it allows the cluster scheduler to optimize the scheduling of your job.
```

> However, sometimes it is beneficial to execute multiple operations within one cluster job,
> like if the compute cluster can only make reservation for full nodes or if there is a limit to the number of cluster jobs you can submit to the cluster scheduler's queue.

Reviewer 3: ```suggestion
However, sometimes it is beneficial to execute multiple operations within one cluster job, like if the compute cluster can only make reservation for full nodes or if there is a limit to the number of cluster jobs you can submit to the cluster scheduler's queue.
```

> .. _cluster_submission_directives:

Reviewer 3: This anchor should be immediately above the header `Submission directives`. Please move it back down so that links continue to point to the right place.

Author: oops

You can execute multiple job-operations in serial within one cluster submission with the ``--bundle`` option.

> For example, the following command will bundle up to five job-operations to be executed into a single cluster submission:

Reviewer 3: We might need to clarify that this will submit multiple cluster jobs, each containing up to five signac-flow job-operations. e.g. 12 job-operations will produce two submissions of 5 and one submission of 2 (remainder).
```suggestion
For example, the following command will bundle up to five job-operations in each cluster submission:
```

Author: Good idea. I also changed the previous sentence to use "per cluster submission".

.. code-block:: bash

    ~/my_project $ python project.py submit --bundle=5

Without any argument, the ``--bundle`` option will bundle **all** eligible job-operations into a single cluster job.

.. tip::

    Recognizing that ``--bundle=1`` is the default option might help you to better understand the bundling concept.

> By default, the submit command will run bundled job-operations in serial. It is possible to run multiple CPU processes in parallel, as long as the operating system assigns running threads to any available CPU cores, which is not available on all compute clusters.

Reviewer 3: Line break and a slight edit here: "which is not the default behavior on all compute clusters."
```suggestion
By default, the submit command will run bundled job-operations in serial.
It is possible to run multiple CPU processes in parallel, as long as the operating system assigns running threads to any available CPU cores, which is not the default behavior on all compute clusters.
```

The ``--parallel`` flag will start the bundled job-operations and run them as background processes. Ensure that the processes are correctly assigned to the requested resources before using this option.

.. warning::

>    The ``--parallel`` flag will not distribute operations among multiple GPUs. Use aggregation instead. (TODO: link to tutorial)

Reviewer 3: ```suggestion
  The ``--parallel`` flag will not distribute operations among multiple GPUs.
  To distribute across multiple GPUs, clusters may require a split MPI communicator.
  An example of this behavior can be found in the `signac-examples project flow.aggregation-mpi <https://github.com/glotzerlab/signac-examples/tree/master/projects/flow.aggregation-mpi>`__.
```

Reviewer 4: I had a distribution issue with CPU as well. It'd be good to mention that there's a possibility that distribution in multiple CPUs could also not work

Author: Does the in-line comment address this about not being default on all systems?

Reviewer 3: @syjlee is right. `--bundle --parallel` may not work on CPUs either. The symptoms are similar: in the GPU case, all the launched "parallel" jobs bind to the first GPU. In the CPU case, all the launched "parallel" jobs bind to the first CPU core.

Binding across multiple GPUs with `--bundle --parallel` is _never_ supported and aggregation is required. The situation with CPUs depends on the cluster's configurations for CPU scheduling within a node.

Reviewer 4: @bdice thanks for crystal clear explanation. For parallel jobs in cpu, I was able to use `--bundle` in Expanse, but not in Bridges 2. So definitely system by system with aforementioned reason.
